### PR TITLE
add R.thrush

### PR DESCRIPTION
**Pull request by guillaumearm**
from #1846,
R.thrush is the T combinator.
examples : http://goo.gl/8tp3jI
with lenses : http://goo.gl/RgLubh

### CODE REVIEW

**Comment by guillaumearm** [comment]
maybe we should wait R.T name be available according to the issue #1846

File: src/thrush.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ *        firstName: "Bruce",
+ *        lastName: "Wayne",
+ *      };
+ *      var tState = R.thrush(state);
+ *      var lastName = lensProp('lastName');
+ *
+ *      tState(R.view(lastName)); //=> "Wayne"
+ *
+ *      var overState = R.compose(tState, over);
+ *      overState(lastName, R.toUpper); //=> {"firstName": "Hello", "lastName": "WORLD"}
```
Comment (CrossEye) [inline src/thrush.js]: I think the lens examples are overkill here.  Someone wanting to understand `thrush` suddenly has to know how `lensProp`, `view`, and `over` all work.

**Comment by CrossEye** [comment]
I like having the function, but as much as I like Smullyan, I don't agree with using the bird names for our functions.  We should have a more descriptive name.  It's too bad that `apply` is already taken.  I don't have a great one now.  `perform` is the best I have to offer.  But I really think it should be something other than `T` or `thrush`, although I would like notes about those alternative names.

**Comment by guillaumearm** [comment]
I'm totally agree with you, i think `perform` is a good idea ;)
anyway, the exemple is overkill, i think i will rewrite one soon.

**Comment by davidchambers** [comment]
I don't like the name `perform`. It sounds very side-effecty. We could instead:

  - remove `call` and `apply` and name the T combinator `apply`; or
  - name the T combinator `apply1`.

**Comment by CrossEye** [comment]
> I don't like the name `perform`. It sounds very side-effecty. 

Hmm, I don't get that impression at all.  But if you do, probably others would as well.

> We could instead:
>   - remove `call` and `apply` and name the T combinator `apply`; or

While I have no love for `call` -- it seems very rarely useful -- I use `apply` all the time, converting external variadic functions into ones that accept a list instead.  Things like `minimum = R.apply(Math.min)`.

>   - name the T combinator `apply1`.

That sounds like a last resort.  If we can't find anything else that would suit, I would live with that, but we've managed to avoid that sort of wonkiness so far.  It would be nice to continue to do so.

**Comment by guillaumearm** [comment]
About `perform`, I think everyone should now Ramda is free-side-effects, so the confusion doesn't really exist ?

and what about `transform` ?

**Comment by guillaumearm** [comment]
simple example.
```
const transform = x => f => f(x)

const state = {
  firstName: 'Hello',
  lastName: 'World',
}

transform(state)(pipe(
  toPairs,
  flatten,
))
```

**Comment by buzzdecafe** [comment]
`exec` anyone? too java-ish?

**Comment by kedashoe** [comment]
> While I have no love for call -- it seems very rarely useful -- I use apply all the time, converting external variadic functions into ones that accept a list instead.

I also don't use call but use apply quite often. How about

apply -> spread
call -> deprecate/leave it alone
thrush -> apply

although there is also unapply.. unspread?

**Comment by guillaumearm** [comment]
Sure, apply is a good name for T, but what is about backward compatibility ? will there be a Ramda 1.0 ?

**Comment by buzzdecafe** [comment]
are we just stuck on the name? if we can't agree on a name but like the function, let's merge and change the name later

**Comment by guillaumearm** [comment]
@buzzdecafe : I agree with you, but not `R.thrush` ?
and I will update the documentation example with the simple example. are you ok for this ?

**Comment by CrossEye** [comment]
But bikeshedding is such an important part of our culture, @buzzdecafe!   Are you trying to spoil our fun?

**Comment by guillaumearm** [comment]
I know ! 

why not `R.with` ?

```
with (41) (add(1)) //=> 42
```

**Comment by buzzdecafe** [comment]
let's just merge as `thrush` FFS.

**Comment by CrossEye** [comment]
I say merge as any of`perform`, `transform`, or `exec`, not `thrush`.

**Comment by itaied246** [comment]
Any plans on merging it?

**Comment by buzzdecafe** [comment]
i think the only hangup is the name. So let's see if we can get one more round of bikeshedding:

`applyTo` anyone?

If we can't come to some agreement, i'd be happy to merge as is

:cow2:

**Comment by itaied246** [comment]
I find `pipeFirst` very informative and clear, 
but I would use anything just merge this PR.

**Comment by kedashoe** [comment]
I like `applyTo` :+1:

**Comment by buzzdecafe** [comment]
@CrossEye you bless `applyTo`? Let's move this one along

**Comment by CrossEye** [comment]
`applyTo` sounds good to me.

**Comment by kedashoe** [comment]
@guillaumearm looks like the bike is finally in the shed. Could you rebase and call the function `applyTo`?

**Comment by haskellcamargo** [comment]
Yay! Updates on this?

**Pull request closed without merging**